### ISSUE TITLE
`holdSubject` does not works properly when `next` is called from `scan`

### ISSUE THREAD

**Issue by safareli** [mostjs-community/subject#8]
when calling `next` from `scan` weird stuff is happening.
here is simple example reproducing this issue:

``` javascript
var mostSubject = require('most-subject');
var subject = new mostSubject.holdSubject(10);
var acc = subject.stream.scan(function(acc, i) {
  console.log('// '+acc+', '+i);
  var result = acc + i;
  if (i == 3) {
    subject.observer.next(4);
  }
  console.log('\\\\ '+result);
  return result;
}, 0);


var subscription = acc.observe( (x) => { console.log('Next: ' + x) });
subject.observer.next(1);
subject.observer.next(2);
subject.observer.next(3);
setTimeout(subject.observer.complete, 0);
```

here is the log.

```
Next: 0
// 0, 1
\\ 1
Next: 1
// 1, 2
\\ 3
Next: 3
// 3, 3
// 3, 4
\\ 7
Next: 7
\\ 6
Next: 6
// 6, 4
\\ 10
Next: 10
```

if `complete` is not called using `setTimeout`

``` diff
- setTimeout(subject.observer.complete, 0);
+ subject.observer.complete();
```

then this is what log looks like:

```
Next: 0
// 0, 1
\\ 1
Next: 1
// 1, 2
\\ 3
Next: 3
// 3, 3
\\ 6
Next: 6
```

but same code using Rx.ReplaySubject works ok.

``` javascript
var Rx = require('rx');
var subject = new Rx.ReplaySubject(10);
var acc = subject.scan(function(acc, i) {
  console.log('// '+acc+', '+i);
  var result = acc + i;
  if (i == 3) {
    subject.onNext(4);
  }
  console.log('\\\\ '+result);
  return result;
}, 0);


var subscription = acc.subscribe(
  (x)   => { console.log('Next: ' + x); },
  (err) => { console.log('Error: ' + err); }, 
  ()    => { console.log('Completed'); }
);

subject.onNext(1);
subject.onNext(2);
subject.onNext(3);
subject.onCompleted();
```

here is the log:

```
// 0, 1
\\ 1
Next: 1
// 1, 2
\\ 3
Next: 3
// 3, 3
\\ 6
Next: 6
// 6, 4
\\ 10
Next: 10
Completed
```

**Comment by TylorS** [comment]
Hmmmm, sorry for not responding sooner. I'll try to investigate this soon

**Comment by TylorS** [comment]
I'm trying to decide if this *is* a bug. I haven't really sought after to recreate Rx type subjects 100%. Rx.ReplaySubject *will* replay values after it is completed, which I find to be sort of strange behavior. Not doing this is also more in line with the rest of *most*. Anyways, I'll try to explain what happens here.

The *issue* here is that complete() is 100% synchronous, and most.js subscriptions `.observe()/.drain()/.reduce()/.forEach() ` are asynchronous.

I've created a jsbin [here](http://jsbin.com/yazile/edit?js,console)

So what happens

"first tick"
1. The holdSubject is created with a buffer of 10
2. The scan is declared
3. The (async) subscription is declared
4. '1' is pushed to the stream queue (sync)
5. '2' is pushed to the stream queue (sync)
6. '3' is pushed to the stream queue (sync)
7. an 'end' event is pushed to the stream queue (sync)
8. The next tick occurs - subscription 'activates' the stream
9. The queue drains the `1`, `2`, `3`, and `end` events


Basically the stream receives the `end` event and is queued before the `4` event. IOW, [this.active](https://github.com/TylorS/most-subject/blob/master/src/Observer.js#L28) of the observer is already going to be set to `false` before the `4` event so [observer.next(4)](https://github.com/TylorS/most-subject/blob/master/src/Observer.js#L37) is essentially a no-op.

Does any of that make sense?

**Comment by TylorS** [comment]
So using the `setTimeout(observer.complete)` allows for the `observer.next(4)` to occur before ending the stream.

**Comment by safareli** [comment]
yes it's now clear about  setTimeout. but what my main issue was that i had stream of some user actions and i was using `.scan` on that with some initial value to generate some state and at some point I may push action to stream from inside of scan and did not want it to be lost. with Rx it's possible to do so by using ReplaySubject or `shareReplay` method on stream [example of shareReplay](https://jsbin.com/dadacixova/edit?js,console)

**Comment by TylorS** [comment]
@safareli I think this can be solved using v5.0.0 :)
It only took forever, but there are asynchronously emitting subjects now which remove the need for `setTimeout`

**Comment by TylorS** [comment]
I'm going to close this issue as it's been 2 months since there was any activity.
Feel free to open this issue back up or open a new one to track any other issues.

Thanks!
